Add Unified ParT to tau NanoAOD

lucasrussell01 opened the pull request:
#### PR description:

This PR adds Unified ParT information to taus stored in NanoAOD.
-> These changes require the PRs from @AlexDeMoor to be merged https://github.com/cms-sw/cmssw/pull/44641, based on the model in https://github.com/cms-data/RecoBTag-Combined/pull/57

Added (for PUPPI jets):
- UnifiedParT discriminators vs jets, electrons and muons,
- Decay mode corresponding with the UnifiedParT tau node with the highest score,
- Correction for tau pt,
- Charge assignment confidence (signed),
- UnifiedParT scores for individual decay-modes (normalised to the sum of the scores for all tau nodes).

Corresponding ParticleNet 2023 variables are stored for CHS jets.

Performance of UParT is described in: https://indico.cern.ch/event/1403350/contributions/5899772/attachments/2832957/4949984/UParT%202024%20-%20final%20results.pdf

Backport PR is here: https://github.com/cms-sw/cmssw/pull/44723

#### PR validation:

Tested successfully with NanoAOD workflow using CMSSW_14_0 RELVAL file, with pending PR UnifiedParT changes and model merged.

Distributions of UParT and PNet for taus compared here:
https://indico.cern.ch/event/1401414/contributions/5891063/attachments/2838625/4961188/UParT_Integration.pdf

lucasrussell01 committed:
Add unified Tagger variables (CHS and PUPPI)

lucasrussell01 committed:
Update framework for Unified Tagging on CHS jets

lucasrussell01 committed:
Update DQM plotting for UTag

lucasrussell01 committed:
update discriminator configuration

lucasrussell01 committed:
first attempt to read in UParT scores

lucasrussell01 committed:
UParT scores added to hybrid producer

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"414e2b7179f7e065b893122d2a4d107ca28d91e9":{"files":["PhysicsTools/NanoAOD/python/nano_cff.py","PhysicsTools/NanoAOD/python/taus_cff.py","PhysicsTools/PatAlgos/plugins/PATTauHybridProducer.cc"],"squashed":false,"time":1712661373},"887a733ffc33e93b89bd8255e01abce259f9d023":{"files":["PhysicsTools/NanoAOD/python/nano_cff.py","PhysicsTools/PatAlgos/plugins/PATTauHybridProducer.cc"],"squashed":false,"time":1712747364},"a09afe5e3626c4c07f58c8e652ab5c4871f5f100":{"files":["PhysicsTools/NanoAOD/python/nano_cff.py","PhysicsTools/NanoAOD/python/taus_cff.py","PhysicsTools/PatAlgos/plugins/PATTauHybridProducer.cc"],"squashed":false,"time":1712661405},"b29db9d7b35fcb81e0f9d2116b13c9d0a9d9f500":{"files":["PhysicsTools/NanoAOD/python/nanoDQM_cfi.py"],"squashed":false,"time":1712661387},"d35e48d69066a5397c3f842114d6052bb8c184bf":{"files":["PhysicsTools/NanoAOD/python/nano_cff.py","PhysicsTools/NanoAOD/python/taus_cff.py","PhysicsTools/PatAlgos/plugins/PATTauHybridProducer.cc"],"squashed":false,"time":1712661080},"f2969bd9d4759da1bf64ff554f93793a58a53555":{"files":["PhysicsTools/NanoAOD/python/nano_cff.py"],"squashed":false,"time":1712747343}},"emoji":{},"last_seen_sha":"887a733ffc33e93b89bd8255e01abce259f9d023","signatures":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39893

- This PR adds an extra 44KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39893/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39893/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

lucasrussell01 committed:
Update DQM Plots for UParT

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39894

- This PR adds an extra 48KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39894/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39894/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

lucasrussell01 committed:
UParT Node Selection Formatting

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39897

- This PR adds an extra 52KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

cmsbuild commented:
A new Pull Request was created by  @lucasrussell01 for master.

It involves the following packages:

- PhysicsTools/NanoAOD (**xpog**)
- PhysicsTools/PatAlgos (**reconstruction, xpog**)


@mandrenguyen, @hqucms, @jfernan2, @vlimant, @cmsbuild can you please review it and eventually sign? Thanks.
@mmarionncern, @emilbols, @AlexDeMoor, @mariadalfonso, @hatakeyamak, @demuller, @jdolen, @jdamgov, @schoef, @AnnikaStein, @Senphy, @ahinzmann, @mbluj, @Ming-Yan, @gkasieczka, @rappoccio, @azotz, @JyothsnaKomaragiri, @andrzejnovak, @nhanvtran, @gouskos, @seemasharmafnal, @gpetruc this is something you requested to watch as well.
@sextonkennedy, @antoniovilela, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

lucasrussell01 committed:
Option to rerun UParT (if not in miniAOD)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39902

- This PR adds an extra 48KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

cmsbuild commented:
Pull request #44691 was updated. @mandrenguyen, @vlimant, @hqucms, @jfernan2, @cmsbuild can you please check and sign again.

lucasrussell01 committed:
add PUPPI jet four vector

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39913

- This PR adds an extra 52KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

cmsbuild commented:
Pull request #44691 was updated. @mandrenguyen, @jfernan2, @hqucms, @cmsbuild, @vlimant can you please check and sign again.

lucasrussell01 committed:
adjust pT correction for recovered taus

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39928

- This PR adds an extra 52KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39928/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39928/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

lucasrussell01 committed:
disable adding of UParT scores for Run 2

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39929

- This PR adds an extra 48KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

cmsbuild commented:
Pull request #44691 was updated. @jfernan2, @mandrenguyen, @vlimant, @cmsbuild, @hqucms can you please check and sign again.

lucasrussell01 committed:
Disable UParT for Run 2 and early Run 3

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39930

- This PR adds an extra 216KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #44641
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #44637, #44641, #44685
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575, #44685

cmsbuild commented:
Pull request #44691 was updated. @mandrenguyen, @cmsbuild, @hqucms, @vlimant, @jfernan2 can you please check and sign again.

mandrenguyen commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/summary.html
**COMMIT**: d30f793080da8c83538fff946135dfe61cd20daf
**CMSSW**: CMSSW_14_1_X_2024-04-13-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/44691/38830/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test runtestPhysicsToolsPatAlgos had ERRORS
</pre>

## RelVals

- [140.023](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/140.023_RunZeroBias2022B)```140.023_RunZeroBias2022B/step3_RunZeroBias2022B.log```
- [140.043](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/140.043_RunZeroBias2022C)```140.043_RunZeroBias2022C/step3_RunZeroBias2022C.log```
- [135.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/135.4_ZEEFS_13)```135.4_ZEEFS_13/step4_ZEEFS_13.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [140.063](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/140.063_RunZeroBias2022D)
- [141.042](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/141.042_RunZeroBias2023D)
- [141.044](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/141.044_RunJetMET2023D)
- [141.046](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/141.046_RunEGamma2023D)
- [1330.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/1330.0_ZMM_13)
- [13234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/13234.0_TTbar_14TeV+2021FS)
- [14034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/14034.0_TTbar_14TeV+2023FS)
- [14234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/14234.0_TTbar_14TeV+2023FSPU)
- [11634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/11634.0_TTbar_14TeV+2021)
- [12434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/12434.0_TTbar_14TeV+2023)
- [12834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/12834.0_TTbar_14TeV+2024)
- [2500.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/2500.4_NANOmc132X)
- [12846.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/12846.0_ZEE_14+2024)
- [12834.7](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/12834.7_TTbar_14TeV+2024_trackingMkFit)
- [25202.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/25202.0_TTbar_13)
- [13034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/13034.0_TTbar_14TeV+2024PU)
- [10224.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrix-results/10224.0_TTbar_13+2017PU)
</details>



## RelVals-INPUT

- [136.72412](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/136.72412_RunJetHT2016B_reminiaodUL)```136.72412_RunJetHT2016B_reminiaodUL/step2_RunJetHT2016B_reminiaodUL.log```
- [140.202](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/140.202_RunJetMET2022D_reMINI)```140.202_RunJetMET2022D_reMINI/step2_RunJetMET2022D_reMINI.log```
- [13234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/13234.0_TTbar_14TeV+2021FS)```13234.0_TTbar_14TeV+2021FS/step2_TTbar_14TeV+2021FS.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [13434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/13434.0_TTbar_14TeV+2021FSPU)
- [14034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/14034.0_TTbar_14TeV+2023FS)
- [14234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/14234.0_TTbar_14TeV+2023FSPU)
- [2500.3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.3_NANOmc130X)
- [2500.301](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.301_EGMNANOmc130X)
- [2500.32](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.32_NANOdata130Xrun3)
- [2500.321](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.321_NANOdata130Xrun3)
- [2500.322](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.322_muPOGNANO130Xrun3)
- [2500.324](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.324_EGMNANOdata130Xrun3)
- [2500.325](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.325_BTVNANO_data13.0)
- [2500.326](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.326_jmeNANOdata130Xrun3)
- [2500.403](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.403_BTVNANO_mc13.2)
- [2500.401](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.401_muPOGNANOmc132X)
- [2500.327](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.327_lepTrackInfoNANOdata130Xrun3)
- [2500.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.4_NANOmc132X)
- [2500.404](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.404_jmeNANOmc132X)
- [2500.402](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.402_EGMNANOmc132X)
- [2500.405](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38830/runTheMatrixINPUT-results/2500.405_lepTrackInfoNANOmc132X)
</details>

lucasrussell01 commented:
Hi @mandrenguyen @hqucms @vlimant @jfernan2, just to quickly discuss these failed tests. 

It seems that the RelVal errors are related to the fact that testing needs to be done with the changes in https://github.com/cms-sw/cmssw/pull/44641 merged - as otherwise RecoBTag.ONNXRuntime.pfUnifiedParticleTransformerAK4JetTags_cfi doesn't exist. 
Also am I correct in thinking the the failed unit test is a general problem, and not related to this PR?

Please let me know if you need anything from our side to move forward :)

hqucms commented on PhysicsTools/NanoAOD/python/nano_cff.py:
```diff
@@ -138,16 +138,30 @@ def nanoAOD_addBoostedTauIds(process, idsToRun=[]):
+        
+        if usePUPPIjets: # option to use PUPPI jets   
+            jetCollection = "updatedJetsPuppi"
+            TagName = "pfUnifiedParticleTransformerAK4JetTag"
```
It seems that the code always uses `TagName+"s"` -- then why not just changing `TagName` itself to have the `s`?

lucasrussell01 replied:
This is how things had been setup for the previous PR that added PNet scores - but indeed, I don't see a reason not to add the `s`

hqucms commented on PhysicsTools/NanoAOD/python/nano_cff.py:
```diff
@@ -138,16 +138,30 @@ def nanoAOD_addBoostedTauIds(process, idsToRun=[]):
+            tag_prefix = "byUTagPUPPI"
+            updatedTauName = originalTauName+'WithUTagPUPPI'
+            # Unified ParT Tagger used for PUPPI jets
+            from RecoBTag.ONNXRuntime.pfUnifiedParticleTransformerAK4JetTags_cfi import pfUnifiedParticleTransformerAK4JetTags
+            process.pfUnifiedParticleTransformerAK4JetTags = pfUnifiedParticleTransformerAK4JetTags.clone()
```
```suggestion
            from RecoBTag.ONNXRuntime.pfUnifiedParticleTransformerAK4_cff import pfUnifiedParticleTransformerAK4JetTags
            Discriminators = pfUnifiedParticleTransformerAK4JetTags.flav_names.value()
```

- Please avoid attaching `pfUnifiedParticleTransformerAK4JetTags` to `process` as you do not intend to run it as a module.

lucasrussell01 replied:
Sure, will implement in an alternative way.

hqucms commented on PhysicsTools/NanoAOD/python/nano_cff.py:
```diff
@@ -138,16 +138,30 @@ def nanoAOD_addBoostedTauIds(process, idsToRun=[]):
+            # PNet tagger used for CHS jets
+            process.load('RecoBTag.ONNXRuntime.pfParticleNetFromMiniAODAK4_cff')
+              
+        Discriminators = [];
+        for tag in getattr(process,TagName+"s").flav_names.value():
+            Discriminators.append(TagName+"s:"+tag)
```
```suggestion
```

hqucms commented on PhysicsTools/NanoAOD/python/nano_cff.py:
```diff
@@ -138,16 +138,30 @@ def nanoAOD_addBoostedTauIds(process, idsToRun=[]):
+            tag_prefix = "byUTagCHS"
+            updatedTauName = originalTauName+'WithUTagCHS'
+            # PNet tagger used for CHS jets
+            process.load('RecoBTag.ONNXRuntime.pfParticleNetFromMiniAODAK4_cff')
```
```suggestion
            from RecoBTag.ONNXRuntime.pfParticleNetFromMiniAODAK4_cff import pfParticleNetFromMiniAODAK4CHSCentralJetTags
            Discriminators = pfParticleNetFromMiniAODAK4CHSCentralJetTags.flav_names.value()
```

lucasrussell01 replied:
Thanks, will implement

lucasrussell01 replied:
In the end for this, we didn't go for this exact form (we decided to add the tag name in front of the discriminator names for a safety/cross check in our producer) - but I used imports instead of adding to process as requested

hqucms commented on PhysicsTools/NanoAOD/python/nano_cff.py:
```diff
@@ -248,12 +272,29 @@ def nanoAOD_customizeCommon(process):
+                            addUTagInfo = nanoAOD_tau_switch.addPNet.value(),
+                            runUTagAK4 = nanoAOD_tau_switch.runUTagPUPPIAK4.value(),
+                            usePUPPIjets = True
+        )
```
I don't think the current implementation of `nanoAOD_addUTagToTaus()` allows adding both the CHS-based PNet and the Puppi-based UParT to the tau collection -- the second call would simply override the first call... 

Please clarify what is the intended behavior of the code and rework `nanoAOD_addUTagToTaus` if needed.

hqucms commented:
enable nano

hqucms commented:
test parameters:
  - pull_requests: https://github.com/cms-data/RecoBTag-Combined/pull/57, https://github.com/cms-sw/cmssw/pull/44641

hqucms commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38853/summary.html
**COMMIT**: d30f793080da8c83538fff946135dfe61cd20daf
**CMSSW**: CMSSW_14_1_X_2024-04-15-1100/el8_amd64_gcc12
**Additional Tests**: NANO
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44691/38853/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @hjbossi cms-sw/cmssw#44549
 - @kyungminparkdrums cms-sw/cmssw#44674

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38853/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38853/git-merge-result


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test runtestPhysicsToolsPatAlgos had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 16 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-15-1100+73a430/62356/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 945 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-15-1100+73a430/62356/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3319227
 - DQMHistoTests: [Total failures: 217](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-15-1100+73a430/62356/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3318990
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 45.55500000000001 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 11634.0,... ): 3.894 KiB        Physics/NanoAODDQM
 - DQMHistoSizes: changed ( 13234.0,... ): 2.033 KiB        Physics/NanoAODDQM
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-15-1100+73a430/62356/triggerResults): no differences found


## NANO Comparison Summary



**Summary**:
 - You potentially [added 48 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-04-15-1100+73a430/62351/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 63 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-04-15-1100+73a430/62351/validateJR.html)
 - DQMHistoTests: Total files compared: 15
 - DQMHistoTests: Total histograms compared: 16456
 - DQMHistoTests: [Total failures: 35](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-04-15-1100+73a430/62351/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 16421
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 41.489 KiB( 14 files compared)
 - DQMHistoSizes: changed ( 2500.001,... ): 3.894 KiB        Physics/NanoAODDQM
 - DQMHistoSizes: changed ( 2500.011,... ): 2.033 KiB        Physics/NanoAODDQM
 - Checked 55 log files, 32 edm output root files, 15 DQM output files

**Nano size comparison Summary**:

| Sample | kb/ev | ref kb/ev | diff kb/ev | ev/s/thd | ref ev/s/thd | diff rate | mem/thd | ref mem/thd |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 2500.0 | 2.783 | 2.695 | 0.088 ( +3.3% ) | 3.39 | 4.24 | -20.1% | 2.190 | 2.205 |
| 2500.001 | 2.897 | 2.805 | 0.092 ( +3.3% ) | 3.04 | 3.77 | -19.2% | 2.615 | 2.676 |
| 2500.002 | 2.844 | 2.753 | 0.090 ( +3.3% ) | 3.12 | 3.95 | -21.1% | 2.655 | 2.658 |
| 2500.01 | 1.447 | 1.386 | 0.061 ( +4.4% ) | 5.64 | 7.32 | -22.9% | 2.339 | 2.309 |
| 2500.011 | 1.907 | 1.823 | 0.083 ( +4.6% ) | 3.18 | 4.28 | -25.8% | 2.520 | 2.528 |
| 2500.012 | 1.761 | 1.687 | 0.074 ( +4.4% ) | 4.50 | 5.62 | -20.0% | 2.394 | 2.447 |
| 2500.1 | 2.354 | 2.339 | 0.015 ( +0.6% ) | 4.44 | 4.50 | -1.3% | 1.981 | 2.071 |
| 2500.2 | 2.459 | 2.443 | 0.015 ( +0.6% ) | 5.09 | 5.08 | +0.2% | 1.899 | 1.983 |
| 2500.21 | 1.286 | 1.274 | 0.012 ( +0.9% ) | 3.40 | 3.49 | -2.4% | 2.172 | 2.343 |
| 2500.211 | 1.668 | 1.653 | 0.015 ( +0.9% ) | 2.77 | 2.86 | -3.0% | 2.187 | 2.370 |
| 2500.3 | 2.229 | 2.214 | 0.015 ( +0.7% ) | 9.47 | 9.19 | +3.0% | 1.962 | 1.970 |
| 2500.301 | 2.833 | 2.815 | 0.019 ( +0.7% ) | 5.96 | 6.69 | -10.9% | 1.820 | 1.956 |
| 2500.31 | 7.164 | 7.164 | 0.000 ( +0.0% ) | 0.77 | 0.96 | -20.0% | 1.649 | 1.702 |
| 2500.311 | 1.568 | 1.568 | 0.000 ( +0.0% ) | 2.47 | 3.13 | -21.1% | 1.052 | 1.053 |
| 2500.312 | 540.457 | 540.457 | 0.000 ( +0.0% ) | 0.41 | 0.43 | -4.4% | 1.520 | 1.584 |
| 2500.313 | 817.694 | 817.694 | 0.000 ( +0.0% ) | 0.53 | 0.55 | -4.2% | 1.516 | 1.567 |
| 2500.32 | 1.350 | 1.354 | -0.004 ( -0.3% ) | 12.18 | 12.42 | -1.9% | 2.240 | 2.377 |
| 2500.321 | 1.759 | 1.761 | -0.002 ( -0.1% ) | 6.30 | 6.82 | -7.6% | 2.272 | 2.334 |
| 2500.322 | 1.240 | 1.240 | 0.000 ( +0.0% ) | 6.52 | 7.23 | -9.8% | 2.100 | 2.167 |
| 2500.323 | 7.772 | 7.772 | 0.000 ( +0.0% ) | 1.07 | 1.22 | -12.6% | 1.777 | 1.902 |
| 2500.324 | 1.880 | 1.882 | -0.002 ( -0.1% ) | 6.29 | 6.86 | -8.4% | 2.257 | 2.313 |
| 2500.325 | 4.167 | 4.291 | -0.124 ( -2.9% ) | 3.65 | 3.64 | +0.1% | 2.176 | 2.243 |
| 2500.326 | 3.336 | 3.205 | 0.131 ( +4.1% ) | 1.53 | 1.70 | -9.9% | 2.115 | 2.238 |
| 2500.327 | 1.815 | 1.816 | -0.001 ( -0.1% ) | 6.61 | 7.86 | -15.9% | 2.145 | 2.289 |
| 2500.4 | 2.374 | 2.388 | -0.014 ( -0.6% ) | 8.74 | 9.16 | -4.6% | 1.821 | 1.838 |
| 2500.401 | 1.891 | 1.891 | 0.000 ( +0.0% ) | 6.23 | 7.01 | -11.1% | 1.709 | 1.819 |
| 2500.402 | 2.950 | 2.962 | -0.012 ( -0.4% ) | 5.19 | 6.88 | -24.7% | 1.739 | 1.876 |
| 2500.403 | 8.700 | 8.918 | -0.219 ( -2.5% ) | 2.54 | 2.63 | -3.2% | 1.754 | 1.794 |
| 2500.404 | 5.473 | 5.272 | 0.202 ( +3.8% ) | 1.13 | 1.27 | -10.7% | 1.831 | 1.809 |
| 2500.405 | 2.860 | 2.872 | -0.012 ( -0.4% ) | 5.17 | 6.96 | -25.7% | 1.837 | 1.894 |
| 2500.5 | 5.194 | 5.194 | 0.000 ( +0.0% ) | 14.82 | 15.39 | -3.7% | 1.524 | 1.570 |
| 2500.51 | 9.120 | 9.120 | 0.000 ( +0.0% ) | 9.31 | 9.61 | -3.1% | 1.516 | 1.517 |

lucasrussell01 commented on PhysicsTools/NanoAOD/python/nano_cff.py:
```diff
@@ -168,13 +183,20 @@ def nanoAOD_addPNetToTaus(process, addPNetInfo=False, runPNetCHSAK4=False):
+        elif runUTagAK4 and usePUPPIjets:
+            # Add Unified ParT tagger information if not available in miniAOD
+            from PhysicsTools.NanoAOD.jetsAK4_Puppi_cff import nanoAOD_addDeepInfoAK4
+            process = nanoAOD_addDeepInfoAK4(process,
```
Ok - will look into this, thanks

lucasrussell01 commented on PhysicsTools/NanoAOD/python/taus_cff.py:
```diff
@@ -20,12 +20,12 @@
 finalTaus = cms.EDFilter("PATTauRefSelector",
     src = cms.InputTag("slimmedTaus"),
-    cut = cms.string("pt > 18 && ((tauID('decayModeFindingNewDMs') > 0.5 && (tauID('byLooseCombinedIsolationDeltaBetaCorr3Hits') || (tauID('chargedIsoPtSumdR03')+max(0.,tauID('neutralIsoPtSumdR03')-0.072*tauID('puCorrPtSum'))<2.5) || tauID('byVVVLooseDeepTau2017v2p1VSjet') || tauID('byVVVLooseDeepTau2018v2p5VSjet'))) || (?isTauIDAvailable('byPNetVSjetraw')?tauID('byPNetVSjetraw'):-1) > {})".forma ...
+    cut = cms.string("pt > 18 && ((tauID('decayModeFindingNewDMs') > 0.5 && (tauID('byLooseCombinedIsolationDeltaBetaCorr3Hits') || (tauID('chargedIsoPtSumdR03')+max(0.,tauID('neutralIsoPtSumdR03')-0.072*tauID('puCorrPtSum'))<2.5) || tauID('byVVVLooseDeepTau2017v2p1VSjet') || tauID('byVVVLooseDeepTau2018v2p5VSjet'))) || (?isTauIDAvailable('byUTagCHSVSjetraw')?tauID('byUTagCHSVSjetraw'):-1) > {} | ...
```
A simple study was performed by @danielwinterbottom - He used a ttbar sample, and tried to find balance between the amount of non_HPS taus not matched to genuine ones and efficiency of all genuine taus (HPS and non-HPS) passing the cut. 0.05 was deemed to be a reasonable threshold.

hqucms commented:
@lucasrussell01 Thanks a lot for addressing the comments! 
Could you please add to the PR description some more info about the validations performed for this PR, e.g. any performance/discriminator distribution plots/slides? And the setups used to derive these plots?

lucasrussell01 commented:
> @lucasrussell01 Thanks a lot for addressing the comments! Could you please add to the PR description some more info about the validations performed for this PR, e.g. any performance/discriminator distribution plots/slides? And the setups used to derive these plots?

I'll push the changes addressing the comments in the next couple of hours. In terms of the discriminator performance for taus, some more information is available here: https://indico.cern.ch/event/1403350/contributions/5899772/attachments/2832957/4949984/UParT%202024%20-%20final%20results.pdf

lucasrussell01 committed:
Tagger Discriminator Names Updated

hqucms commented:
> > @lucasrussell01 Thanks a lot for addressing the comments! Could you please add to the PR description some more info about the validations performed for this PR, e.g. any performance/discriminator distribution plots/slides? And the setups used to derive these plots?
> 
> @hqucms I'll push the changes addressing the comments in the next couple of hours. In terms of the discriminator performance for taus, some more information is available here: https://indico.cern.ch/event/1403350/contributions/5899772/attachments/2832957/4949984/UParT%202024%20-%20final%20results.pdf

Thank you @lucasrussell01 !

For the performance, I was more looking for some studies performed within the TAU POG, particularly with the implementations in this PR. Could you or the @cms-sw/tau-pog-l2 point us to them?

lucasrussell01 committed:
Modify addition of Deep Info

lucasrussell01 commented:
> Thank you @lucasrussell01 !
> 
> For the performance, I was more looking for some studies performed within the TAU POG, particularly with the implementations in this PR. Could you or the @cms-sw/tau-pog-l2 point them to us?

@hqucms What in particular would you like to see? 

We checked the nano outputs visually, can produce some histograms of the added variables in Nano if you like.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44691/39963

- This PR adds an extra 116KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/taus_cff.py  modified in PR(s): #44575

cmsbuild commented:
Pull request #44691 was updated. @jfernan2, @vlimant, @hqucms, @mandrenguyen, @cmsbuild can you please check and sign again.

hqucms commented:
> > Thank you @lucasrussell01 !
> > For the performance, I was more looking for some studies performed within the TAU POG, particularly with the implementations in this PR. Could you or the @cms-sw/tau-pog-l2 point them to us?
> 
> @hqucms What in particular would you like to see?
> 
> We checked the nano outputs visually, can produce some histograms of the added variables in Nano if you like.

OK that would be good. Could you please plot both PNet and UParT so we can have also a qualitative comparison?

lucasrussell01 committed:
Tau switch to add UParT

lucasrussell01 commented:
> > > Thank you @lucasrussell01 !
> > > For the performance, I was more looking for some studies performed within the TAU POG, particularly with the implementations in this PR. Could you or the @cms-sw/tau-pog-l2 point them to us?
> > 
> > 
> > @hqucms What in particular would you like to see?
> > We checked the nano outputs visually, can produce some histograms of the added variables in Nano if you like.
> 
> OK that would be good. Could you please plot both PNet and UParT so we can have also a qualitative comparison?

@hqucms Ok, working on making these now

cmsbuild commented:
Pull request #44691 was updated. @jfernan2, @mandrenguyen, @vlimant, @hqucms, @cmsbuild can you please check and sign again.

hqucms commented:
please test

lucasrussell01 commented:
> > > Thank you @lucasrussell01 !
> > > For the performance, I was more looking for some studies performed within the TAU POG, particularly with the implementations in this PR. Could you or the @cms-sw/tau-pog-l2 point them to us?
> > 
> > 
> > @hqucms What in particular would you like to see?
> > We checked the nano outputs visually, can produce some histograms of the added variables in Nano if you like.
> 
> OK that would be good. Could you please plot both PNet and UParT so we can have also a qualitative comparison?

Hi @hqucms,

Some plots and brief comments are available here: https://indico.cern.ch/event/1401414/contributions/5891063/attachments/2838625/4961188/UParT_Integration.pdf 

I will add this to the PR description.

hqucms commented:
> > > > Thank you @lucasrussell01 !
> > > > For the performance, I was more looking for some studies performed within the TAU POG, particularly with the implementations in this PR. Could you or the @cms-sw/tau-pog-l2 point them to us?
> > > 
> > > 
> > > @hqucms What in particular would you like to see?
> > > We checked the nano outputs visually, can produce some histograms of the added variables in Nano if you like.
> > 
> > 
> > OK that would be good. Could you please plot both PNet and UParT so we can have also a qualitative comparison?
> 
> Hi @hqucms,
> 
> Some plots and brief comments are available here: https://indico.cern.ch/event/1401414/contributions/5891063/attachments/2838625/4961188/UParT_Integration.pdf
> 
> I will add this to the PR description.

Thanks a lot @lucasrussell01 , that's very useful to have!

Can you also update the backport PR so we can start testing it?

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38869/summary.html
**COMMIT**: a02009cdde0e936905cc04726dbbaebd405ef88d
**CMSSW**: CMSSW_14_1_X_2024-04-16-1100/el8_amd64_gcc12
**Additional Tests**: NANO
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44691/38869/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @AdrianoDee cms-sw/cmssw#44742
 - @AlexDeMoor cms-sw/cmssw#44641

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38869/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73a430/38869/git-merge-result


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test runtestPhysicsToolsPatAlgos had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 10 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-16-1100+73a430/62370/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 936 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-16-1100+73a430/62370/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3319227
 - DQMHistoTests: [Total failures: 231](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-16-1100+73a430/62370/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3318976
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 45.55500000000001 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 11634.0,... ): 3.894 KiB        Physics/NanoAODDQM
 - DQMHistoSizes: changed ( 13234.0,... ): 2.033 KiB        Physics/NanoAODDQM
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-16-1100+73a430/62370/triggerResults): no differences found


## NANO Comparison Summary



**Summary**:
 - You potentially [added 5 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-04-16-1100+73a430/62368/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 63 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-04-16-1100+73a430/62368/validateJR.html)
 - DQMHistoTests: Total files compared: 15
 - DQMHistoTests: Total histograms compared: 16456
 - DQMHistoTests: [Total failures: 35](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-04-16-1100+73a430/62368/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 16421
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 41.489 KiB( 14 files compared)
 - DQMHistoSizes: changed ( 2500.001,... ): 3.894 KiB        Physics/NanoAODDQM
 - DQMHistoSizes: changed ( 2500.011,... ): 2.033 KiB        Physics/NanoAODDQM
 - Checked 55 log files, 32 edm output root files, 15 DQM output files

**Nano size comparison Summary**:

| Sample | kb/ev | ref kb/ev | diff kb/ev | ev/s/thd | ref ev/s/thd | diff rate | mem/thd | ref mem/thd |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 2500.0 | 2.783 | 2.695 | 0.088 ( +3.3% ) | 3.52 | 4.43 | -20.4% | 2.252 | 2.237 |
| 2500.001 | 2.897 | 2.805 | 0.092 ( +3.3% ) | 3.17 | 4.00 | -20.6% | 2.776 | 2.650 |
| 2500.002 | 2.843 | 2.753 | 0.090 ( +3.3% ) | 3.24 | 4.08 | -20.5% | 2.671 | 2.648 |
| 2500.01 | 1.446 | 1.386 | 0.061 ( +4.4% ) | 6.06 | 7.99 | -24.1% | 2.315 | 2.272 |
| 2500.011 | 1.906 | 1.823 | 0.083 ( +4.5% ) | 3.32 | 4.43 | -25.1% | 2.481 | 2.418 |
| 2500.012 | 1.761 | 1.687 | 0.074 ( +4.4% ) | 4.86 | 6.25 | -22.3% | 2.467 | 2.351 |
| 2500.1 | 2.354 | 2.339 | 0.015 ( +0.6% ) | 4.53 | 4.51 | +0.5% | 2.071 | 2.066 |
| 2500.2 | 2.459 | 2.443 | 0.015 ( +0.6% ) | 5.11 | 5.20 | -1.7% | 1.982 | 1.979 |
| 2500.21 | 1.286 | 1.274 | 0.012 ( +0.9% ) | 3.49 | 3.51 | -0.5% | 2.271 | 2.262 |
| 2500.211 | 1.668 | 1.653 | 0.015 ( +0.9% ) | 3.17 | 3.10 | +2.5% | 2.351 | 2.365 |
| 2500.3 | 2.229 | 2.214 | 0.015 ( +0.7% ) | 9.76 | 9.73 | +0.2% | 1.974 | 1.968 |
| 2500.301 | 2.833 | 2.815 | 0.019 ( +0.7% ) | 8.27 | 8.38 | -1.3% | 1.959 | 1.952 |
| 2500.31 | 7.164 | 7.164 | 0.000 ( +0.0% ) | 1.31 | 1.43 | -8.0% | 1.707 | 1.708 |
| 2500.311 | 1.568 | 1.568 | 0.000 ( +0.0% ) | 6.65 | 6.98 | -4.6% | 1.056 | 1.058 |
| 2500.312 | 540.457 | 540.457 | 0.000 ( +0.0% ) | 0.53 | 0.53 | -0.4% | 1.595 | 1.599 |
| 2500.313 | 817.694 | 817.694 | 0.000 ( +0.0% ) | 0.72 | 0.74 | -2.6% | 1.587 | 1.585 |
| 2500.32 | 1.350 | 1.354 | -0.004 ( -0.3% ) | 12.41 | 12.78 | -2.9% | 2.367 | 2.367 |
| 2500.321 | 1.759 | 1.761 | -0.002 ( -0.1% ) | 8.46 | 8.92 | -5.2% | 2.418 | 2.453 |
| 2500.322 | 1.240 | 1.240 | 0.000 ( +0.0% ) | 9.12 | 9.23 | -1.2% | 2.138 | 2.210 |
| 2500.323 | 7.772 | 7.772 | 0.000 ( +0.0% ) | 3.36 | 3.45 | -2.7% | 1.928 | 1.921 |
| 2500.324 | 1.880 | 1.882 | -0.002 ( -0.1% ) | 9.06 | 9.30 | -2.6% | 2.166 | 2.202 |
| 2500.325 | 4.167 | 4.291 | -0.124 ( -2.9% ) | 4.37 | 4.10 | +6.5% | 2.169 | 2.202 |
| 2500.326 | 3.337 | 3.205 | 0.131 ( +4.1% ) | 1.66 | 1.78 | -6.5% | 2.278 | 1.763 |
| 2500.327 | 1.815 | 1.816 | -0.001 ( -0.1% ) | 9.24 | 9.37 | -1.4% | 2.309 | 2.299 |
| 2500.4 | 2.374 | 2.388 | -0.014 ( -0.6% ) | 9.28 | 9.31 | -0.4% | 1.800 | 1.728 |
| 2500.401 | 1.891 | 1.891 | 0.000 ( +0.0% ) | 8.31 | 8.33 | -0.3% | 1.689 | 1.690 |
| 2500.402 | 2.950 | 2.962 | -0.012 ( -0.4% ) | 8.12 | 8.02 | +1.3% | 1.802 | 1.877 |
| 2500.403 | 8.700 | 8.918 | -0.219 ( -2.5% ) | 3.02 | 2.80 | +7.6% | 1.836 | 1.945 |
| 2500.404 | 5.474 | 5.272 | 0.202 ( +3.8% ) | 1.21 | 1.32 | -7.8% | 1.789 | 1.756 |
| 2500.405 | 2.860 | 2.872 | -0.012 ( -0.4% ) | 8.16 | 8.19 | -0.3% | 1.803 | 1.901 |
| 2500.5 | 5.194 | 5.194 | 0.000 ( +0.0% ) | 15.87 | 15.76 | +0.7% | 1.558 | 1.491 |
| 2500.51 | 9.120 | 9.120 | 0.000 ( +0.0% ) | 9.47 | 9.60 | -1.4% | 1.510 | 1.514 |

hqucms commented:
+1

 - The failed unit test is also in IB and not related to this PR.
 - This PR only changes the `Tau_` branches in NanoAOD. Other changes in Nano/Mini (bDiscriminators) are from the dependency PR https://github.com/cms-sw/cmssw/pull/44641.

mandrenguyen commented:
+reconstruction
This doesn't touch reco, only nanoAOD (despite what tests appear to show), as explained here:
https://github.com/cms-sw/cmssw/pull/44691#issuecomment-2059742085

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @sextonkennedy, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
type tau

hqucms commented:
@cms-sw/orp-l2 Could you please merge this? Thanks!

antoniovilela commented:
Thanks

antoniovilela commented:
externals already merged
